google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/gaxx/retrying/ApiResultRetryAlgorithm.java

```diff
@@ -18,17 +18,46 @@
 import com.google.api.core.InternalApi;
 import com.google.api.gax.retrying.BasicResultRetryAlgorithm;
 import com.google.api.gax.retrying.RetryingContext;
+import com.google.api.gax.retrying.TimedAttemptSettings;
 import com.google.api.gax.rpc.ApiException;
+import com.google.protobuf.util.Durations;
+import com.google.rpc.RetryInfo;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.protobuf.ProtoUtils;
+import org.threeten.bp.Duration;
 
-/** For internal use, public for technical reasons. */
+/**
+ * For internal use, public for technical reasons. This retry algorithm checks the metadata of an
+ * exception for additional error details. If the metadata has a RetryInfo field, use the retry
+ * delay to set the wait time between attempts.
+ */
 @InternalApi
 public class ApiResultRetryAlgorithm<ResponseT> extends BasicResultRetryAlgorithm<ResponseT> {
 
+  private static final Metadata.Key<RetryInfo> KEY_RETRY_INFO =
+      ProtoUtils.keyForProto(RetryInfo.getDefaultInstance());
+
+  @Override
+  public TimedAttemptSettings createNextAttempt(
+      Throwable prevThrowable, ResponseT prevResponse, TimedAttemptSettings prevSettings) {
+    Duration retryDelay = extractRetryDelay(prevThrowable);
+    if (retryDelay != null) {
+      return prevSettings
+          .toBuilder()
+          .setRandomizedRetryDelay(retryDelay)
+          .setAttemptCount(prevSettings.getAttemptCount() + 1)
```

Comment:
I wonder if its possible to align with grpc native retries reset the next non-server instructed delay to the initial?

https://github.com/grpc/proposal/blob/master/A6-client-retries.md#pushback